tutorials/ann_hnswlib.ipynb

```diff
@@ -0,0 +1,394 @@
+    "    user_based=True,\n",
+    ").run()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b00b5236-1a94-43a9-bfb9-6dc4623bb33c",
+   "metadata": {},
+   "source": [
+    "## Building index for ANN recommender\n",
+    "\n",
+    "After MF model is trained, we need to wrap it with an ANN recommender. We employ Cornac built-in HNSWLibANN which implements [HNSW algorithm](https://arxiv.org/abs/1603.09320) for building index and doing approximate K-nearest neighbor search. More on how to tune the hyper-parameters at https://github.com/nmslib/hnswlib."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "id": "fa280b9d-ec04-41eb-9de2-acfb67fbeb80",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ann = HNSWLibANN(\n",
+    "    recom=mf,\n",
+    "    M=16,\n",
+    "    ef_construction=100,\n",
+    "    ef=50,\n",
+    "    num_threads=-1,\n",
+    ")\n",
+    "ann.build_index()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b67fe0eb-7ca5-4ff9-87a5-a34ab612036e",
+   "metadata": {},
+   "source": [
+    "## Time/accuracy tradeoff"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b1846494-e618-48ef-b489-45d02e0d5430",
+   "metadata": {},
+   "source": [
+    "Here we measure the tradeoff between efficiency and accuracy. Let say we do top-20 recommendations for 10,000 users."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "46c5f959-f706-4406-85c9-a4acf2ac20c0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "K = 20\n",
+    "N = 10000\n",
+    "test_users = np.random.choice(mf.user_ids, size=N)\n",
+    "mf_recs = []\n",
+    "ann_recs = []"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "6730a93f-729d-4bea-b5aa-8fb9d9cef867",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 2min 40s, sys: 19.5 ms, total: 2min 40s\n",
+      "Wall time: 5.04 s\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "for uid in test_users:\n",
+    "    mf_recs.append(mf.recommend(uid, k=K))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "520da23e-79eb-4c53-9aaa-a33aaea883fc",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 312 ms, sys: 31 µs, total: 312 ms\n",
+      "Wall time: 309 ms\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "for uid in test_users:\n",
+    "    ann_recs.append(ann.recommend(uid, k=K))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e2f4f68a-c69b-4e70-b32a-81eb78d21279",
+   "metadata": {},
+   "source": [
+    "While it took MF 5.04s to complete the task, it's only 309ms for ANN. The speed up is about 15 times. Note that our dataset contains less than 5000 items. We will see an even bigger improvement with more items and with higher dimensions of the factors."
```

Comment:
with higher dimensions of the factors --> with higher dimensional factors

Reply:
fixed